src/expr.rs

```diff
@@ -889,39 +887,72 @@ impl Rewrite for ast::Arm {
-                                                                        line_indent + 4)) {
-                if first_line_width(body_str) <= budget {
+            let offset = Indent::new(offset.block_indent,
+                                     line_start + 4 - offset.block_indent);
+            let rewrite = nop_block_collapse(body.rewrite(context, budget, offset), budget);
+
+            match rewrite {
+                Some(ref body_str) if body_str.len() <= budget || comma.is_empty() =>
                     return Some(format!("{}{} => {}{}",
                                         attr_str.trim_left(),
                                         pats_str,
                                         body_str,
-                                        comma));
-                }
+                                        comma)),
+                _ => rewrite,
             }
-        }
+        } else {
+            None
+        };
 
-        // We have to push the body to the next line.
-        if comma.is_empty() {
+        if let ast::ExprBlock(_) = body.node {
             // We're trying to fit a block in, but it still failed, give up.
             return None;
         }
 
         let body_budget = try_opt!(width.checked_sub(context.config.tab_spaces));
-        let body_str = try_opt!(nop_block_collapse(body.rewrite(context,
-                                                                body_budget,
-                                                                context.block_indent)));
-        Some(format!("{}{} =>\n{}{},",
+        let next_line_body = nop_block_collapse(body.rewrite(context,
+                                                             body_budget,
+                                                             context.block_indent
+                                                                    .block_indent(context.config)),
+                                                body_budget);
+
+        let (body_str, break_line) = try_opt!(match_arm_heuristic(same_line_body.as_ref()
+                                                                                .map(|x| &x[..]),
+                                                                  next_line_body.as_ref()
+                                                                                .map(|x| &x[..])));
+
+        let spacer = if break_line {
+            format!("\n{}",
+                    offset.block_indent(context.config).to_string(context.config))
+        } else {
+            " ".to_owned()
+        };
+
+        Some(format!("{}{} =>{}{},",
                      attr_str.trim_left(),
                      pats_str,
-                     offset.block_indent(context.config).to_string(context.config),
+                     spacer,
                      body_str))
     }
 }
 
+// Takes two possible rewrites for the match arm body and chooses the "nicest".
+// Bool marks break line or no.
```

Comment:
The bool seems only to indicate if we took the second option, rather than actually checking for line breaks. It would be nice if we didn't need it tbh.